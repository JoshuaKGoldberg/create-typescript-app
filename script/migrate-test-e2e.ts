import chalk from "chalk";
import { $, execaCommand } from "execa";
import * as fs from "node:fs/promises";
import { rimraf } from "rimraf";
import { assert, describe, expect, test } from "vitest";

const filesExpectedToBeChanged = [
	"README.md",
	"knip.json",
	".github/workflows/ci.yml",
	".gitignore",
	".prettierignore",
	"cspell.json",
	"eslint.config.js",
];

const filesThatMightBeChanged = new Set([
	...filesExpectedToBeChanged,
	"script/__snapshots__/migrate-test-e2e.ts.snap",
]);

await rimraf("coverage*");

const originalReadme = (await fs.readFile("README.md")).toString();

const originalSnapshots = (
	await fs.readFile("script/__snapshots__/migrate-test-e2e.ts.snap")
).toString();

await $({
	stdio: "inherit",
})`c8 -o ./coverage -r html -r lcov --src src node ./bin/index.js --auto --mode migrate --skip-all-contributors-api --skip-github-api --skip-install`;

// All Contributors seems to not be using Prettier to format files...
await fs.writeFile(
	".all-contributorsrc",
	JSON.stringify(
		JSON.parse((await fs.readFile(".all-contributorsrc")).toString()),
		null,
		2,
	) + "\n",
);

// Ignore changes to the README.md all-contributor count and contributors table...
const updatedReadme = (await fs.readFile("README.md")).toString();
await fs.writeFile(
	"README.md",
	[
		updatedReadme.slice(0, updatedReadme.indexOf("## Contributors")) +
			originalReadme.slice(
				originalReadme.indexOf("## Contributors"),
				originalReadme.indexOf("<!-- markdownlint-restore -->"),
			),
		updatedReadme.slice(updatedReadme.indexOf("<!-- markdownlint-restore -->")),
	]
		.join("")
		.replace(
			/All Contributors: \d+/g,
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			/All Contributors: \d+/.exec(originalReadme)![0],
		)
		.replace(
			/all_contributors-\d+/g,
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			/all_contributors-\d+/.exec(originalReadme)![0],
		),
);

// ...and even to the snapshot file, so diffs don't mind it.
await fs.writeFile(
	"script/__snapshots__/migrate-test-e2e.ts.snap",
	originalSnapshots
		.replace(
			/All Contributors: \d+/g,
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			/All Contributors: \d+/.exec(originalReadme)![0],
		)
		.replace(
			/all_contributors-\d+/g,
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			/all_contributors-\d+/.exec(originalReadme)![0],
		),
);

describe("expected file changes", () => {
	test.each(filesExpectedToBeChanged)("%s", async (file) => {
		const { stdout } = await execaCommand(`git diff HEAD -- ${file}`);
		const contentsAfterGitMarkers = stdout
			.split("\n")
			.slice(2)
			.join("\n")
			.replace(/@@ -\d+,\d+ \+\d+,\d+ @@/g, "@@ ... @@");

		assert(
			stdout,
			`Looks like there were no changes to ${file} from migration?`,
		);

		// If this fails, see .github/DEVELOPMENT.md > Setup Scripts for context.
		// Then see .github/DEVELOPMENT.md > Migration Snapshot Failures.
		expect(contentsAfterGitMarkers).toMatchSnapshot();
	});
});

test("unexpected file changes", async () => {
	const { stdout: gitStatus } = await $`git status`;
	console.log(`Stdout from running \`git status\`:\n${gitStatus}`);

	const indexOfUnstagedFilesMessage = gitStatus.indexOf(
		"Changes not staged for commit:",
	);
	assert(
		indexOfUnstagedFilesMessage !== -1,
		`Looks like migrate didn't cause any file changes? That's ...probably incorrect? ðŸ˜¬`,
	);

	const unstagedModifiedFiles = gitStatus
		.slice(indexOfUnstagedFilesMessage)
		.match(/modified: {3}\S+\n/g)
		?.map((match) => match.split(/\s+/)[1])
		.filter((filePath) => !filesThatMightBeChanged.has(filePath));

	console.log("Unexpected modified files are:", unstagedModifiedFiles);

	if (unstagedModifiedFiles?.length) {
		const gitDiffCommand = `git diff HEAD -- ${unstagedModifiedFiles.join(
			" ",
		)}`;
		const { stdout } = await execaCommand(gitDiffCommand);

		console.log(`Stdout from running \`${gitDiffCommand}\`:\n${stdout}`);

		throw new Error(
			[
				"",
				"Oh no! Running the migrate script unexpectedly modified:",
				...unstagedModifiedFiles.map((filePath) => ` - ${filePath}`),
				"",
				"See .github/DEVELOPMENT.md > Setup Scripts for context.",
				"Then see .github/DEVELOPMENT.md > Unexpected File Modifications.",
			]
				.map((line) => chalk.red(line))
				.join("\n"),
		);
	}
});
