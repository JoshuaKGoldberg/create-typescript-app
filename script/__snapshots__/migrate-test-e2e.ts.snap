// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expected file changes > .github/workflows/ci.yml 1`] = `
"--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ ... @@ jobs:
       - uses: ./.github/actions/prepare
       - run: pnpm run test --coverage
       - uses: codecov/codecov-action@v3
-        with:
-          flags: unit
   type_check:
     name: Type Check
     runs-on: ubuntu-latest"
`;

exports[`expected file changes > .gitignore 1`] = `
"--- a/.gitignore
+++ b/.gitignore
@@ ... @@
-coverage*/
+coverage/
 lib/
 node_modules/"
`;

exports[`expected file changes > .prettierignore 1`] = `
"--- a/.prettierignore
+++ b/.prettierignore
@@ ... @@
 .all-contributorsrc
 .husky/
-coverage*/
+coverage/
 lib/
 pnpm-lock.yaml"
`;

exports[`expected file changes > README.md 1`] = `
"--- a/README.md
+++ b/README.md
@@ ... @@ Thanks! ðŸ’–
 
 <!-- ALL-CONTRIBUTORS-LIST:END -->
 <!-- spellchecker: enable -->
+
+<!-- You can remove this notice if you don't want it ðŸ™‚ no worries! -->
+
+> ðŸ’™ This package was templated with [\`create-typescript-app\`](https://github.com/JoshuaKGoldberg/create-typescript-app)."
`;

exports[`expected file changes > cspell.json 1`] = `
"--- a/cspell.json
+++ b/cspell.json
@@ ... @@
 {
 	"dictionaries": ["typescript"],
 	"ignorePaths": [
-		"./coverage*",
-		"./script/__snapshots__",
 		".github",
 		"CHANGELOG.md",
+		"coverage",
 		"lib",
 		"node_modules",
 		"pnpm-lock.yaml"
@@ ... @@
 		"apexskier",
 		"arethetypeswrong",
 		"automerge",
-		"codecov",
 		"codespace",
 		"contributorsrc",
 		"execa",
 		"infile",
+		"joshuakgoldberg",
 		"knip",
+		"markdownlint",
 		"markdownlintignore",
 		"mtfoley",
 		"npmignore",
 		"npmpackagejsonlintrc",
 		"outro",
 		"packagejson",
-		"precommit",
 		"quickstart",
 		"tada",
 		"tseslint",
 		"tsup",
-		"vitest"
+		"vitest",
+		"wontfix"
 	]
 }"
`;

exports[`expected file changes > eslint.config.js 1`] = `
"--- a/eslint.config.js
+++ b/eslint.config.js
@@ ... @@
-/*
-ðŸ‘‹ Hi! This ESLint configuration contains a lot more stuff than many repos'!
-You can read from it to see all sorts of linting goodness, but don't worry -
-it's not something you need to exhaustively understand immediately. ðŸ’™
-
-If you're interested in learning more, see the 'getting started' docs on:
-- ESLint: https://eslint.org
-- typescript-eslint: https://typescript-eslint.io
-*/
-
 import eslint from "@eslint/js";
 import comments from "@eslint-community/eslint-plugin-eslint-comments/configs";
 import vitest from "@vitest/eslint-plugin";
@@ ... @@ export default tseslint.config(
 		languageOptions: {
 			parserOptions: {
 				projectService: {
-					allowDefaultProject: ["*.config.*s", "bin/*.js", "script/*.ts"],
+					allowDefaultProject: ["*.*s", "eslint.config.js"],
 					defaultProject: "./tsconfig.json",
 				},
 				tsconfigRootDir: import.meta.dirname,
@@ ... @@ export default tseslint.config(
 			"no-constant-condition": "off",
 
 			// These on-by-default rules work well for this repo if configured
-			"@typescript-eslint/no-unnecessary-condition": [
-				"error",
-				{
-					allowConstantLoopConditions: true,
-				},
-			],
 			"@typescript-eslint/no-unused-vars": ["error", { caughtErrors: "all" }],
-			"@typescript-eslint/prefer-nullish-coalescing": [
-				"error",
-				{ ignorePrimitives: true },
-			],
-			"@typescript-eslint/restrict-template-expressions": [
-				"error",
-				{ allowBoolean: true, allowNullish: true, allowNumber: true },
-			],
 			"n/no-unsupported-features/node-builtins": [
 				"error",
 				{ allowExperimental: true },"
`;

exports[`expected file changes > knip.json 1`] = `
"--- a/knip.json
+++ b/knip.json
@@ ... @@
 {
 	"$schema": "https://unpkg.com/knip@latest/schema.json",
-	"entry": ["src/index.ts!", "script/*e2e.js"],
+	"entry": ["src/index.ts!"],
 	"ignoreExportsUsedInFile": { "interface": true, "type": true },
-	"project": ["src/**/*.ts!", "script/**/*.js"]
+	"project": ["src/**/*.ts!"]
 }"
`;
